File: concrete/src/main/java/pcb/uwu/amount/quantity/Grams.java
package pcb.uwu.amount.quantity;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.quantity.MassUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.quantity.GramUnit.GRAM;

public class Grams extends Mass {

	// region constructors

	public Grams(Number value) {
		super(value, GRAM);
	}

	public Grams(Number value, Magnitude magnitude) {
		super(value, magnitude, GRAM);
	}

	public Grams(String value) {
		super(value, GRAM);
	}

	public Grams(String value, Magnitude magnitude) {
		super(value, magnitude, GRAM);
	}

	public Grams(BigDecimal value) {
		super(value, GRAM);
	}

	public Grams(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, GRAM);
	}

	public Grams(BigDecimalAmount amount) {
		super(amount, GRAM);
	}

	public Grams(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, GRAM);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Grams plus(UnitAmount<MassUnit> other, MathContext mathContext) {
		return new Grams(plusAmount(other, mathContext));
	}

	@Override
	public Grams minus(UnitAmount<MassUnit> other, MathContext mathContext) {
		return new Grams(minusAmount(other, mathContext));
	}

	@Override
	public Grams multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Grams(multipliedByScalar(other, mathContext));
	}

	@Override
	public Grams dividedBy(BigDecimal other, MathContext mathContext) {
		return new Grams(dividedByScalar(other, mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Grams.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.GramUnit
import pcb.uwu.unit.base.MassUnit

class Grams : Mass
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, GramUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, GramUnit)

    // region UnitAmount

    override fun plus(mass: UnitAmount<MassUnit>) =
        Grams(this.amount + (mass to GramUnit).amount)

    override fun minus(mass: UnitAmount<MassUnit>) =
        Grams(this.amount - (mass to GramUnit).amount)

    override fun times(scalar: Number) =
        Grams(this.amount * scalar)

    override fun div(scalar: Number) =
        Grams(this.amount / scalar)

    // endregion
}
